### DISCUSSION TITLE
Render Texture using FBO

### DISCUSSION THREAD

**Discussion by Nyubug** [bnpr/Malt#562] [Q&A] [answered]
Hello! 

So, my goal is to create a material that can do the following. Render a camera view and specify the collection of objects into an image in real time and display it on a plane object. The idea is to mimic 2D animation parallax effects by creating different layers such as a mountain and forest. 

The Portal BSDF node sounds perfect on paper, but the problem is the node is limited to Cycles, doesn't have a way to isolate objects (unless we use geo nodes), and doesn't seem to have transparency allowing the portals to overlap each other properly. 

For example.
![Untitledz](https://github.com/bnpr/Malt/assets/162280580/a4d01beb-0edf-4861-8bb2-62ab576175fa)

So, I ended up making a python script that does the following. Sets the view layer of the isolated collection, draws the camera view into an offscreen, read the offscreen texture color, then set the pixels of an image.
<img width="1280" alt="Capture3" src="https://github.com/bnpr/Malt/assets/162280580/ff5b2553-82a4-454c-b991-31163f966dea">

This is the result of the python script. Each collection is rendered into its own image and displayed on a plane object. This is the result I'm aiming for. Animation playback is around 4 FPS for 7 images each at 1024 pixels which is very problematic. Technically the final render will be around 512 pixels because we are dividing the camera FOV in half to add overscan for parallax effects.
<img width="1280" alt="Capture2" src="https://github.com/bnpr/Malt/assets/162280580/4ba37171-95c9-4985-9a8e-2d42a7aa566a">

So those were my failed attempts, but I'm curious if this can be done in Malt with better performance. I found this website mentioning FBO and was wondering if this is possible to implement on a mesh material? Or perhaps there is another method and I'm just overthinking it.
https://www.songho.ca/opengl/gl_fbo.html

**Comment by pragma37** [comment]
You can certainly do this by rendering each layer to its own texture.

However if instead of that you scale down the mesh in the axis you want to flatten the object, and scale back the position used for shading, you get the same result with much less trouble (and much faster).

![imagen](https://github.com/bnpr/Malt/assets/36610999/cf85daa9-b0dc-4ad9-b5b4-e00c4c7e8ff4)

![imagen](https://github.com/bnpr/Malt/assets/36610999/94b94434-6e6e-410a-8d4b-8de942708189)

Here I'm skipping the scaling in the shadow pass, so shadows still work normally.
[malt-parallax.zip](https://github.com/user-attachments/files/15920233/malt-parallax.zip)

Note that there was a bug in the Transform node (actually I think is a driver bug) that I just fixed. So download the latest Malt release before opening the file.

**Reply by Nyubug** [reply]
Ah, that is so neat! So, you can retain the shaders by just undoing any vertex transformations and apply the result to the pixel pass?

Its funny you mention flattening the vertices along a given axis but that simply isn't enough in my case. I've experimented with many methods that attempt to "capture" the camera view using geometry nodes but I ultimately gave up because I realized shaders wouldn't work.

My geometry node setup converts the vertices into camera space and then perform a perspective divide using the z depth. This allows the resulting geometry appears as if its perspective when viewed from an orthographic camera. It also allows the resulting geometry to reflect camera movement, geometry changes, etc. Here is my basic node setup to accomplish this.
<img width="1183" alt="Capture7" src="https://github.com/bnpr/Malt/assets/162280580/fdddb0d5-f9fc-49b9-927e-931ebbbaead6">

The reason why I want to use an orthographic camera is so I can transform the geometry location and scale without distortion you get with a perspective camera. This means I can do 2D parallax effects by using an empty objects Z depth to determine the parallax speed of each object. 
<img width="1280" alt="Capture6" src="https://github.com/bnpr/Malt/assets/162280580/1dd3aa33-159a-4b4b-964c-00e7e75a6c44">


Here is a small example of the system in geometry nodes. 
[Flatten_Geo_Node.zip](https://github.com/user-attachments/files/15935864/Flatten_Geo_Node.zip)

EDIT:
I ended up porting my geo nodes over to malt nodes and it works surprisingly well! It's not done yet, but I have a few questions if you don't mind. [malt-parallax-v2.zip](https://github.com/user-attachments/files/15936391/malt-parallax-v2.zip)

- Is there a node similar to the geometry node "Object Info" that allows me to access an object's location and rotation? I'd like to use this material on all objects but each object I can assign a different camera for example. 

- Is there a way to preserve the original vertex location after we've deformed the vertices? For example, If I transform the vertices 10 units away, can I somehow keep the original vertex/shader location visible as well? Or would I have to create a duplicate via geo nodes? 

Honestly Malt is amazing so far and I haven't even explored it fully. Thanks for the fast reply!

**Reply by pragma37** [reply]
> Is there a node similar to the geometry node "Object Info" that allows me to access an object's location and rotation?

There's a node literally called "Object Info" that exposes this. 😅
It's in the "Input" section.

> Is there a way to preserve the original vertex location after we've deformed the vertices?

While you could make a shader plugin where you declare your own "varying" variable and functions to set/get it, it's easier to just store it in a vertex color if you're not using the 4 vertex color slots.
There's no built-in node to create or override arrays, but you can work around it by declaring it with the `Inline Code` node:
`vec4[4]( COLOR[0], COLOR[1], COLOR[2], vec4(a, 1) )`

![imagen](https://github.com/bnpr/Malt/assets/36610999/d4b1d630-9b17-415a-9aef-a7455b90d44b)

**Reply by Nyubug** [reply]
Hmm.. I did see the Object Info node, but I didn't see an option to specify an object which led me to believe it only works on the object that is using the material. In geo nodes there is an input allowing you to choose an object. 
I'm not familiar with how Malt works yet, It's a bit of a learning curve. 😐

**Reply by pragma37** [reply]
Ah, I misunderstood what you meant. Yes, the Malt's Object Info is limited to the object being rendered.
For other objects you would have to use drivers.

**Reply by Nyubug** [reply]
So even though I can preserve shaders after deforming the vertices by storing the vertex positions, how can I preserve other things such as shadows?

I found out that if I don't flatten some of the Suzanne's, the shadows are saved on the floor. But the moment I apply the vertex transformation to the Suzanne's the shadows are removed from the floor.

There could be other issues such as reflections(normals), bloom, line rendering, ambient occlusion, etc all of which might look different after transforming the vertices. 

So, my question is how can I be sure that everything remains the same after transforming the vertices?
I still think the render texture method would be the best combined with the scene filter node to filter out collections into the texture. I just need a way to expose the camera view as a mesh shader node color output but I'm unsure where to begin.

https://github.com/bnpr/Malt/assets/162280580/3fe8f44e-734d-44d2-a45d-074d32597a7c

**Reply by pragma37** [reply, marked as answer]
> So even though I can preserve shaders after deforming the vertices by storing the vertex positions, how can I preserve other things such as shadows?

On top of not flattening the mesh in shadow maps, you would have to displace it to its "original" position.
Speculars are view depended so you would have to compute them manually too.

> I still think the render texture method would be the best combined with the scene filter node to filter out collections into the texture. I just need a way to expose the camera view as a mesh shader node color output but I'm unsure where to begin.

If you want to take that route, you can use this as a reference since it's somewhat similar:
https://github.com/bnpr/Malt/discussions/559#discussioncomment-9783773
